### DISCUSSION TITLE
How to create and save my own graph data?

### DISCUSSION THREAD

**Discussion by hkim716** [pyg-team/pytorch_geometric#2129] [Q&A]
I'm trying to create my own graph datasets for a GNN model using PyG. 
When I load `citeseer` dataset, I can see the information as `Data(edge_index=[2, 9104], test_mask=[3327], train_mask=[3327], val_mask=[3327], x=[3327, 3703], y=[3327])`. And there were `processed` and `raw` folders that contain several files. e.g. `ind.citeseer.allx`, `ind.citeseer.graph`, and so on.

Suppose that I would like to make graph datasets, with 100 nodes and 3 features at each node. I have feature and node connection information in the form of `numpy array` which is `npz` file. 
How can I efficiently make my own datasets that are compatible with PyTorch Geometric? Is it possible to make the graph datasets using `networkx`, and do you recommend it? Then how can I save the graph datasets such that we could use `.npz` file for image data?

**Comment by rusty1s** [comment]
I do not recommend using `networkx` if your data is already present in the form of `numpy` arrays.
In `dataset.process`, you will need to take care of converting those `numpy` files to the PyG `data` object though. Then, your dataset may look like something like this:
```python
class MyDataset(InMemoryDataset):
    def __init__(self, root, raw_filename, transform)
        self.raw_filename = raw_filename
        super().__init__(root, transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def processed_file_names(self):
        return 'data.pt'

    def process(self):
        data_list = []  # Read raw files and create data list out of them.
        torch.save(self.collate(data_list), self.processed_paths[0])
```

**Reply by hkim716** [reply]
Thanks for your answer Matt, but it's still confusing. Can you give me some detailed code that read `numpy` data?
I would like to make graph datasets with 100 nodes, and each node has only two edges to make a graph the shape of circle. 
Each node has 3 features and there will be no edge features. I have `mydata.npz` dataset, which has all of feature information as a `numpy array`.

**Reply by rusty1s** [reply]
How does `mydata.npz` look like? How is your data encoded?

**Reply by hkim716** [reply]
[mydata.zip](https://github.com/rusty1s/pytorch_geometric/files/6013936/mydata.zip)

I attached `mydata.npz` here. 
I would like to make `class MyDataset(InMemoryDataset)` that prepares `PyG` data.
`mydata.npz` data contains `50 graphs`, and each graph has `100 nodes`, and each node has `x`, `y`, `z` as node features. It also has `edge_index` in the form of `numpy array`.

**Reply by rusty1s** [reply]
That would look like the following:
```python
import torch
import numpy as np
from torch_geometric.data import InMemoryDataset, Data


class MyDataset(InMemoryDataset):
    def __init__(self, root, transform=None):
        super(MyDataset, self).__init__(root, transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return "mydata.npz"

    @property
    def processed_file_names(self):
        return "data.pt"

    def process(self):
        raw_data = np.load(self.raw_paths[0])

        x = torch.from_numpy(raw_data['x'])
        y = torch.from_numpy(raw_data['y'])
        z = torch.from_numpy(raw_data['z'])
        pos = torch.stack([x, y, z], dim=-1)

        edge_index = torch.from_numpy(raw_data['edge_index'])
        edge_index -= 1
        edge_index = edge_index.t().contiguous()

        data_list = []
        for i in range(pos.size(0)):
            data = Data(pos=pos[i], edge_index=edge_index)
            data_list.append(data)

        torch.save(self.collate(data_list), self.processed_paths[0])
```
where `mydata.npz` is saved in a `{root}/raw` directory.

**Reply by hkim716** [reply]
Thanks Matt! It worked! Now I can see `data.pt`, `pre_filter.pt`, `pre_transform.pt` files in `processed` folder. 
I'm trying to build `MyModel(nn.Module)` using the data. 
Firstly, I would like to use just one graph, `data = dataset[0]` to do `node level regression`.
I have stacked GCN layers from `PyG library`, but I got stuck again. 

Here is my code. 

```
import torch
import torch.nn as nn
import torch.nn.functional as F

import torch_geometric.nn as pyg_nn
import torch_geometric.utils as pyg_utils

import numpy as np
import torch
import torch.optim as optim

from MyDatasetClass import MyDataset
from torch_geometric.data import DataLoader

import torch_geometric.transforms as T
import matplotlib.pyplot as plt
import numpy as np



class MyModel(nn.Module):
    def __init__(self, in_channels, out_channels):
        super(MyModel, self).__init__()
        self.conv1 = pyg_nn.GCNConv(in_channels, 64, cached=True)
        self.conv2 = pyg_nn.GCNConv(64, out_channels, cached=True)
        self.conv3 = pyg_nn.GCNConv(out_channels, 64, cached=True)
        self.conv4 = pyg_nn.GCNConv(64, in_channels, cached=True)

    def forward(self, data):
        
        x, edge_index = data.pos, data.edge_index
        x = x.float()
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = self.conv2(x, edge_index)
        emb = x
        x = F.relu(x)
        x = self.conv3(x, edge_index)
        x = F.relu(x)
        x = self.conv4(x, edge_index)

        return emb, torch.mean(x, axis=0)

    def loss(self, pred, label):
        return F.mse_loss(pred, label.to(torch.float32))



dev = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

dataset = MyDataset(root='./')
data = dataset[0]
data = data.to(dev)

def train(dataset):
	loader = DataLoader(dataset, batch_size=1, shuffle=True)
	model = MyModel(3, 16).to(dev)
   	opt = optim.Adam(model.parameters(), lr=0.01)

   	# train
    for epoch in range(200):
        total_loss = 0
        model.train()
        for batch in loader:
            opt.zero_grad()
            embedding, pred = model(batch.to(dev))
            label = batch.pos
            pred = pred[batch.pos]
            label = label[batch.pos]
            loss = model.loss(pred, label)
            loss.backward()
            opt.step()
            total_loss += loss.item() * batch.num_graphs
            
        total_loss /= len(loader.dataset)

        if epoch % 10 == 0:
            test_acc = test(test_loader, model)
            print("Epoch {}. Loss: {:.4f}. Test accuracy: {:.4f}".format(
                epoch, total_loss, test_acc))

    return model



def test(loader, model, is_validation=False):
    model.eval()

    correct = 0
    for data in loader:
        with torch.no_grad():
            emb, pred = model(data.to(dev))
            label = data.pos
        
        mask = data.pos.to(dev) if is_validation else data.pos.to(dev)

        pred = pred[mask]
        label = data.y[mask]
            
        correct += pred.eq(label).sum().item()
    
    total = 0
    for data in loader.dataset:
        total += torch.sum(data.test_mask).item()
    return correct / total

task = 'node'

model = train(dataset)
```

I also did not understand how to define test function `def test` for the test. 
```
---> 35             pred = pred[batch.pos]
     36             label = label[batch.pos]

IndexError: tensors used as indices must be long, byte or bool tensors

```
I got an error like this. I know more things look stupid to you, but I beg your help to make this work. 
I really thank you again!

**Reply by rusty1s** [reply]
1. Please remove the `cached=True` option in `GCNConv` as this will result in a crash when you try to train on multiple graphs.
2. I'm not really sure what you are trying to achieve. It seems like you want to train a global model output against individual input node positions. Note that indexing with `pos` does not make any sense, since you can only index with long tensors. In case you simply want to reconstruct input node features, you can do:
```python
embedding, pred = model(batch.to(dev))
loss = model.loss(pred, label)
```

as they do not make sense since you cannot index with floating point tensors. It is sufficient to

**Reply by hkim716** [reply]
Thanks for your answer. I fixed the wrong indexing with `batch.pos`, which does not make sense.

1. I think I should have made `train_mask` and `test_mask` so I can train/test node features, right?
2. I am curious about `batch_size`. 
```
def train(dataset):
	loader = DataLoader(dataset, batch_size=16, shuffle=True)
```
Based on my dataset, does the `batch_size` mean "the number of `graphs` that I want to group as a batch"?
because when I `print(batch.num_graphs)`, I saw `16 16 16 2` for each epoch, which is same as the total number of graphs, `50` .

**Reply by rusty1s** [reply]
1. You can either split each graph into training and testing nodes via masking, or use unseen graphs as test examples. You can achieve that by splitting your dataset, e.g., via:
```python
train_dataset = dataset[:40]
test_dataset = dataset[40:]
```
2. Yes, `batch_size` does relate to the number of graph/examples you want to group to a mini-batch.

**Reply by hkim716** [reply]
Thanks a lot Matt. I'm getting comfortable with PyG and graph neural networks! 
I'm really interested playing with your awesome libraries, and I would like to go one step further. 
If I want to do graph level predictions (either classification or regression) based on `mydata`, I will need to have `pooling layers` to reduce the number of nodes at each pooling steps, right?
In this case, how would you place `pooling layers` in `MyModel`, which was defined as 

```
class MyModel(nn.Module):
    def __init__(self, in_channels, out_channels):
        super(MyModel, self).__init__()
        self.conv1 = pyg_nn.GCNConv(in_channels, 64)
        self.conv2 = pyg_nn.GCNConv(64, out_channels)
        self.conv3 = pyg_nn.GCNConv(out_channels, 64)
        self.conv4 = pyg_nn.GCNConv(64, in_channels)

    def forward(self, data):
        
        x, edge_index = data.pos, data.edge_index
        x = x.float()
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = self.conv2(x, edge_index)
        emb = x
        x = F.relu(x)
        x = self.conv3(x, edge_index)
        x = F.relu(x)
        x = self.conv4(x, edge_index)

        return emb, torch.mean(x, axis=0)

    def loss(self, pred, label):
        return F.mse_loss(pred, label.to(torch.float32))
```

Thanks in advance Matt!

**Reply by rusty1s** [reply]
In general, global readout functions perform mostly just as well as sophisticated pooling layers, which coarsen the graph after each iteration. The majority of GNN implementations just perform global pooling rather than iterative pooling. You currently already do this via `torch.mean(...)`, but in order to work on mini-batches of `batch_size > 1`, you will need to replace that call with `torch_geometric.nn.global_mean_pool`.

**Reply by hkim716** [reply]
Hey Matt, 
I would like to reduce the dimensionality to `10 nodes` using pooling layer. 
I added a `global_mean_pool`, but I cannot work with it. 
Why I get an error `NameError: name 'batch' is not defined`
Can you fix my code?

```
class MyModel(nn.Module):
    def __init__(self, in_channels, out_channels):
        super(MyModel, self).__init__()
        self.conv1 = pyg_nn.GCNConv(in_channels, 3)
        self.conv2 = pyg_nn.GCNConv(3, out_channels)
        self.conv3 = pyg_nn.GCNConv(out_channels, 3)
        self.pool3 = pyg_nn.global_mean_pool(3, batch, 10)
        self.conv4 = pyg_nn.GCNConv(3, in_channels)
        
  
    def forward(self, data):
        
        x, edge_index, batch = data.pos, data.edge_index, data.batch
        x = x.float()
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = self.conv2(x, edge_index)
        emb = x
        x = F.relu(x)
        x = self.conv3(x, edge_index)
        x = F.relu(x)
        x = pool3(x)
        x = self.conv4(x, edge_index)

        return emb, x

    def loss(self, pred, label):

        return F.mse_loss(pred, label.to(torch.float32))
```

**Reply by rusty1s** [reply]
The `global_mean_pool` method is not a module, but a function. You can simply call it inside your `forward` function:
```python
class MyModel(nn.Module):
    def __init__(self, in_channels, out_channels):
        super(MyModel, self).__init__()
        self.conv1 = pyg_nn.GCNConv(in_channels, 3)
        self.conv2 = pyg_nn.GCNConv(3, out_channels)
        self.conv3 = pyg_nn.GCNConv(out_channels, in_channels)
        
  
    def forward(self, data):
        x, edge_index, batch = data.pos, data.edge_index, data.batch
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = self.conv2(x, edge_index)
        emb = x
        x = F.relu(x)
        x = self.conv3(x, edge_index)
        x = pyg.nn.global_mean_pool(x, batch)

        return emb, x
```

**Reply by hkim716** [reply]
What is the `shape of dataset` from `global_mean_pool`? If I want to reduce my `100 nodes` to `30 nodes`, how can I define `global_mean_pool`?

**Reply by rusty1s** [reply]
`global_mean_pool` will aggregate all node features into a single vector representation. In case you work on a single graph, it will output a `[1, num_features]` matrix. In case you work on multiple graphs, it will output a `[batch_size, num_features]` matrix.

If you want to slowly coarsen your graph, e.g., going from 100 to 30 nodes, I suggest you to look into the `top-k` pooling approaches provided by PyG.
